### ISSUE TITLE
Narrow `match` value type based on earlier cases

### ISSUE THREAD

**Issue by connorbrinton** [python/mypy#20189]
**Bug Report**

Mypy does not exclude types from earlier match cases in subsequent cases. This makes it difficult to work with values that could match multiple cases.

I'm specifically running into this when trying to handle types from the OpenAI and Anthropic Python SDKs. The type annotations in those libraries frequently have patterns like `content: Union[str, Iterable[ContentArrayOfContentPart], None]`. Applying a `match/case` statement to the `content` causes issues in the `case Iterable():` case, because Mypy thinks that the value could still be a string (even though that was handled in a previous case).

**To Reproduce**

[Playground URL](https://mypy-play.net/?mypy=latest&python=3.12&gist=520de783347cc3de504cea3b2a76b6ab)

```python
import random
from collections.abc import Iterable
from typing import assert_never

x: str | list[int]
if random.random() < 0.5:
    x = "a"
else:
    x = [1, 2, 3]

match x:
    case str():
        reveal_type(x)
    case Iterable():
        reveal_type(x)
    case _:
        assert_never(x)
```

**Expected Behavior**

I would expect Mypy to show a revealed type of _only_ `builtins.list[builtins.int]` for `x` on line 17, like:
```
main.py:15: note: Revealed type is "builtins.str"
main.py:17: note: Revealed type is "builtins.list[builtins.int]"
Success: no issues found in 1 source file
```

**Actual Behavior**

Even though a string would never make it to the second case, `x` on line 17 is still considered to possibly be a string:
```
main.py:15: note: Revealed type is "builtins.str"
main.py:17: note: Revealed type is "builtins.str | builtins.list[builtins.int]"
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.12

**Comment by connorbrinton** [comment]
I've been experimenting more with this, and it seems like this is the case even for `if/elif` blocks:
```python
import random
from collections.abc import Iterable
from typing import assert_never

x: str | list[int]
if random.random() < 0.5:
    x = "a"
else:
    x = [1, 2, 3]

if isinstance(x, str):
    reveal_type(x)
elif isinstance(x, Iterable):
    reveal_type(x)
else:
    assert_never(x)
```

Output:
```
test.py:14: note: Revealed type is "builtins.str"
test.py:16: note: Revealed type is "builtins.str | builtins.list[builtins.int]"
Success: no issues found in 1 source file
```

I would guess that there is almost certainly already an issue for `if/elif/else` type narrowing based on earlier conditions, but I haven't been able to find it in my searching 😞

**Comment by hauntsaninja** [comment]
It's specific to the Protocol narrowing. Good news is it is fixed on master by https://github.com/python/mypy/pull/19471 , I guess wait for 1.19 to be released!

**Comment by connorbrinton** [comment]
Brilliant! Thanks for pointing that out, I'll wait for 1.19 to be released! 😄
